### ISSUE TITLE
[Bug]: Error "undefined" launching browser process when using `node:18-buster` Docker image and installing dependencies listed on website

### ISSUE THREAD

**Issue by mattwelke** [puppeteer/puppeteer#10285]
### Bug expectation

**Summary:**

I expected Node.js to be able to execute a script that uses Puppeteer to read an element from a website, with the script being based on the example from the website:

```js
// Import puppeteer
import puppeteer from 'puppeteer';

(async () => {
    // Launch the browser
    const browser = await puppeteer.launch({
        headless: 'new',
    });

    // Create a page
    const page = await browser.newPage();

    // Go to your site
    await page.goto('https://pptr.dev/guides/query-selectors');

    // Query for an element handle.
    const inner_html = await page.$eval('h1', element => element.innerHTML);

    console.log(`Inner HTML: ${inner_html}`);

    // Close browser.
    await browser.close();
})();
```

Instead, Node.js encountered an error launching the browser process:

```
file:///usr/src/app/node_modules/@puppeteer/browsers/lib/esm/launch.js:250
                reject(new Error([
                       ^

Error: Failed to launch the browser process! undefined
[18:18:0531/203130.811200:ERROR:zygote_host_impl_linux.cc(100)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.


TROUBLESHOOTING: https://pptr.dev/troubleshooting

    at ChildProcess.onClose (file:///usr/src/app/node_modules/@puppeteer/browsers/lib/esm/launch.js:250:24)
    at ChildProcess.emit (node:events:525:35)
    at ChildProcess._handle.onexit (node:internal/child_process:291:12)

Node.js v18.16.0
```

Notably, this error occurred even after I tried installing the list of dependencies on the Puppeteer website *and* after I tried installing the list of dependencies on the Chromium website.

**Details (with reproduction example):**

I'm running into a problem getting this to work with the latest version of the `node:18-buster` Docker image.

I have a Node.js script that works if I run it locally without using Docker. It's a slightly modified version of the one I found in the website "getting started" content.

```js
// Import puppeteer
import puppeteer from 'puppeteer';

(async () => {
    // Launch the browser
    const browser = await puppeteer.launch({
        headless: 'new',
    });

    // Create a page
    const page = await browser.newPage();

    // Go to your site
    await page.goto('https://pptr.dev/guides/query-selectors');

    // Query for an element handle.
    const inner_html = await page.$eval('h1', element => element.innerHTML);

    console.log(`Inner HTML: ${inner_html}`);

    // Close browser.
    await browser.close();
})();
```

```
$ node index.mjs 
Inner HTML: Query Selectors
```

(this output is good, because that's what the website has right now)

Then, I tried to Dockerize it because my ultimate goal is to create a runtime for a FaaS provider that would allow users to use Puppeteer from their code. I started with a simple Dockerfile that I expected would not work because I'm not installing anything to support Puppeteer.

```Dockerfile
FROM node:18-buster

WORKDIR /usr/src/app

COPY package*.json ./

RUN npm install

COPY . .

CMD [ "node", "index.mjs" ]
```

I got an error related to starting the process, like I expected.

```
Error: Failed to launch the browser process!
/root/.cache/puppeteer/chrome/linux-113.0.5672.63/chrome-linux64/chrome: error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory


TROUBLESHOOTING: https://pptr.dev/troubleshooting
```

I read that web page, got to [the UNIX instructions](https://pptr.dev/troubleshooting#chrome-headless-doesnt-launch-on-unix) specifically, and made a new Dockerfile that installs those dependencies.

```Dockerfile
FROM node:18-buster

# Install system packages for puppeteer NPM dependency
RUN apt-get -y update \
  && apt-get -y install --no-install-recommends \
  ca-certificates \
  fonts-liberation \
  libasound2 \
  libatk-bridge2.0-0 \
  libatk1.0-0 \
  libc6 \
  libcairo2 \
  libcups2 \
  libdbus-1-3 \
  libexpat1 \
  libfontconfig1 \
  libgbm1 \
  libgcc1 \
  libglib2.0-0 \
  libgtk-3-0 \
  libnspr4 \
  libnss3 \
  libpango-1.0-0 \
  libpangocairo-1.0-0 \
  libstdc++6 \
  libx11-6 \
  libx11-xcb1 \
  libxcb1 \
  libxcomposite1 \
  libxcursor1 \
  libxdamage1 \
  libxext6 \
  libxfixes3 \
  libxi6 \
  libxrandr2 \
  libxrender1 \
  libxss1 \
  libxtst6 \
  lsb-release \
  wget \
  xdg-utils \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*

WORKDIR /usr/src/app

COPY package*.json ./

RUN npm install

COPY . .

CMD [ "node", "index.mjs" ]
```

This didn't work though, which surprised me. It was especially surprising because the error message said the error was "undefined".

```
Error: Failed to launch the browser process! undefined
[18:18:0531/201731.497189:ERROR:zygote_host_impl_linux.cc(100)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.
```

I decided to troubleshoot because the website says:

> Also, see https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/linux/debian/dist_package_versions.json for the up-to-date list of dependencies declared by the Chrome installer.

I followed that link and found a new set of dependencies, listed under "Debian 10 (Buster)". I made a new Dockerfile using them.

```Dockerfile
FROM node:18-buster

# Install system packages for puppeteer NPM dependency
# From source.chromium.org
RUN apt-get update && apt-get install -y --no-install-recommends \
    libasound2=1.1.8-1 \
    libatk-bridge2.0-0=2.30.0-5 \
    libatk1.0-0=2.30.0-2 \
    libatspi2.0-0=2.30.0-7 \
    libc6=2.28-10+deb10u1 \
    libcairo2=1.16.0-4+deb10u1 \
    libcups2=2.2.10-6+deb10u5 \
    libdbus-1-3=1.12.20-0+deb10u1 \
    libdrm2=2.4.97-1 \
    libexpat1=2.2.6-2+deb10u4 \
    libgbm1=18.3.6-2+deb10u1 \
    libgcc1=1:8.3.0-6 \
    libglib2.0-0=2.58.3-2+deb10u3 \
    libnspr4=2:4.20-1 \
    libnss3=2:3.42.1-1+deb10u5 \
    libpango-1.0-0=1.42.4-7~deb10u1 \
    libpangocairo-1.0-0=1.42.4-7~deb10u1 \
    libstdc++6=8.3.0-6 \
    libuuid1=2.33.1-0.1 \
    libx11-6=2:1.6.7-1+deb10u2 \
    libx11-xcb1=2:1.6.7-1+deb10u2 \
    libxcb-dri3-0=1.13.1-2 \
    libxcb1=1.13.1-2 \
    libxcomposite1=1:0.4.4-2 \
    libxcursor1=1:1.1.15-2 \
    libxdamage1=1:1.1.4-3+b3 \
    libxext6=2:1.3.3-1+b2 \
    libxfixes3=1:5.0.3-1 \
    libxi6=2:1.7.9-1 \
    libxkbcommon0=0.8.2-1 \
    libxrandr2=2:1.5.1-1 \
    libxrender1=1:0.9.10-1 \
    libxshmfence1=1.3-1 \
    libxss1=1:1.2.3-1 \
    libxtst6=2:1.2.3-1 \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /usr/src/app

COPY package*.json ./

RUN npm install

COPY . .

CMD [ "node", "index.mjs" ]

```

This time, the Docker image build didn't even complete. Among the errors output by the build were:

```
#0 2.690 E: Version '2.2.10-6+deb10u5' for 'libcups2' was not found
```

I was able to find information about [libcups2](https://packages.debian.org/buster/libcups2). This is when I learned that the most recent version of that package is `2.2.10-6+deb10u6` right now, not `2.2.10-6+deb10u5`. After more troubleshooting, I realized that Debian doesn't seem to publish any version except the most recent version of each package (I tested this with another dependency from that list too).

So then I thought a way to get it working could be to remove the version numbers from that list of dependencies from the Chromium website so that it ends up installing the latest version of each.

```Dockerfile
FROM node:18-buster

# From source.chromium.org, with the specific versions removed.
RUN apt-get update && apt-get install -y --no-install-recommends \
    libasound2 \
    libatk-bridge2.0-0 \
    libatk1.0-0 \
    libatspi2.0-0 \
    libc6 \
    libcairo2 \
    libcups2 \
    libdbus-1-3 \
    libdrm2 \
    libexpat1 \
    libgbm1 \
    libgcc1 \
    libglib2.0-0 \
    libnspr4 \
    libnss3 \
    libpango-1.0-0 \
    libpangocairo-1.0-0 \
    libstdc++6 \
    libuuid1 \
    libx11-6 \
    libx11-xcb1 \
    libxcb-dri3-0 \
    libxcb1 \
    libxcomposite1 \
    libxcursor1 \
    libxdamage1 \
    libxext6 \
    libxfixes3 \
    libxi6 \
    libxkbcommon0 \
    libxrandr2 \
    libxrender1 \
    libxshmfence1 \
    libxss1 \
    libxtst6 \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /usr/src/app

COPY package*.json ./

RUN npm install

COPY . .

CMD [ "node", "index.mjs" ]
```

This didn't work either. It was the same "undefined" error again.

```
Error: Failed to launch the browser process! undefined
[18:18:0531/202504.802808:ERROR:zygote_host_impl_linux.cc(100)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.
```

This concluded my troubleshooting. At this point, I came to the conclusion that either I was doing something wrong or that either the list of required dependencies for Debian on the Puppeteer website or on the Chromium website are out of date.

I think it's also worth noting that there appear to be [other people recently experiencing issues running Puppeteer on Debian too](https://github.com/puppeteer/puppeteer/issues/290).

### Bug behavior

- [ ] Flaky
- [ ] PDF

### Minimal, reproducible example

```TypeScript
Minimal example is multiple files, so see detailed reproducible example above.
```

### Error string

Error: Failed to launch the browser process! undefined

### Puppeteer version

20.5.0

### Node version

18.16.0

### Package manager

npm

### Package manager version

9.6.7

### Operating system

Linux

**Comment by mattwelke** [comment]
Looks like I didn't take that warning that talked about "sandbox" seriously enough.

I was able to search this on the web and ended up getting things to work by changing my Node.js code to launch the browser without the sandbox:

```js
// Launch the browser
const browser = await puppeteer.launch({
    headless: 'new',
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
});
```

Looks like my next steps are to research running as non-root in Docker and what the sandbox mode is all about, and then to decide how to structure my Dockerfile so that things work while still being secure.

**Comment by safety97** [comment]
thank you save my day !

**Comment by DerekGray2022** [comment]
Saved me many days!! Thanks, @mattwelke!!

**Comment by avvRobertoAlma** [comment]
@mattwelke could you please publish your working dockerfile?

**Comment by mattwelke** [comment]
@avvRobertoAlma I don't remember which Dockerfile I ended up using for this but I do remember that things started to work when I fixed my Node.js code. It was my program code that was the problem, not the Dockerfile, because [I was launching Puppeteer in a way that wasn't compatible with the Debian image](https://github.com/puppeteer/puppeteer/issues/10285#issuecomment-1571225635).

Try this:

```Dockerfile
FROM node:18-buster

WORKDIR /usr/src/app

COPY package*.json ./

RUN npm install

COPY . .

CMD [ "node", "index.mjs" ]
```

And if your Node.js code launches Puppeteer correctly, it should work, like it did for me. Then, you'll probably want to double check that `18` is still the most recent Debian version. It might not be now.
